### DISCUSSION TITLE
Node inspector not showing anything when running performance profiler

### DISCUSSION THREAD

**Discussion by osdiab** [vercel/next.js#49837] [Help]
### Summary

Hi, I have run NextJS 13 (pages router) with `NODE_OPTIONS='--inspect'` [as described in these docs](https://nextjs.org/docs/pages/building-your-application/configuring/debugging#debugging-with-chrome-devtools); when opening the Chrome inspector, it shows the dev server in the `Remote Targets` list, and I get a log in my terminal saying `Debugger attached`, so I think that part's working.

But when I go to the `Performance` tab and try running a profile, I only ever get an empty profile that says `(idle)` even if I am doing things that interact with the local dev server. I also get an error in the debugger console saying:

```
Failed to find browser main thread in trace, some timeline features may be unavailable
```

as well as some failures to load sourcemaps. Any ideas what might be going wrong that are preventing us from seeing anything in the profiler?

### Additional information

EDIT: I tried it again now and it looks like I'm getting another useless profile, but with slightly different data:

<img width="790" alt="Screenshot 2023-05-16 at 11 52 57" src="https://github.com/vercel/next.js/assets/1890926/b3bae623-2e7b-46ca-8830-82029cd67f36">

EDIT 2: I tried it again and noticed that when running `next dev`, i'm getting a number of these logs:

```
Starting inspector on 127.0.0.1:9229 failed: address already in use
```

maybe this is relevant? even if i kill the processes that are listening on 9229 the next time I run `next dev` it just spits out these errors again.

**Comment by osdiab** [comment]
OK so I think I'm getting closer. It now appears to be working now that I did the following:

- I made a production build and ran that. this might not be necessary
- I didn't invoke it with `yarn`. I think the duplicate inspector hooks are because I was invoking the script to run the command via `yarn`. Instead used `./node_modules/bin/next`.

I'll try it just with the dev server now.

**Reply by osdiab** [reply]
OK, so dev server doesn't work. It shows the same 

```
Starting inspector on 127.0.0.1:9229 failed: address already in use
```

warning repeatedly. So yeah not sure if this is something peculiar to our NextJS config or what, but basically can only use the inspector meaningfully with the above constraints. weird.

**Reply by andresgutgon** [reply]
Hi @osdiab did you manage to debug with NextJS? I made a sample repo to try to find out what's happening. I started with an standalone nextjs app in javascript with pages directory. 
https://github.com/andresgutgon/nextjs-debugging/tree/main/npm-standalone-nextjs-pages-javascript

I tried these options

### Run NODE_DEV=dev debug from package.json
In the [package.json](https://github.com/andresgutgon/nextjs-debugging/blob/main/npm-standalone-nextjs-pages-javascript/package.json#L7) I used the flag `NODE_OPTIONS='--inspect'`

### Run NODE_ENV=dev debug from terminal
But also tried this:
```
NODE_OPTIONS='--inspect' node_modules/.bin/next dev
```
### Run built next app with `start` command
```
 NODE_OPTIONS='--inspect' node_modules/.bin/next start
```

Neither of these options looks like are working. Any idea?

**Reply by osdiab** [reply]
nope, just ended up doing it as mentioned above, which was not fun!

**Reply by andresgutgon** [reply]
Building and invoking `start` command from `node_modules` didn't work for me neither. Man this shouldn't be so hard if Nextjs want to be seing as a serious backend framework : /

**Reply by osdiab** [reply]
lol i think you're a little late to it being considered serious, given that it's the biggest framework in JS these days. But yeah, surprising that this is a pain point for debugging.

**Reply by andresgutgon** [reply]
I do agree is awesome. But man have being using Rails for 10+ years the dev experience is so much worst when it comes to debugging. 

In rails (or ruby)

``` ruby
def my_thing
  byebug # Server will stop, as simple as that
end
```

**Comment by nikitalarionov** [comment]
Try like that: node inspect ./dist/index.js

**Comment by andresgutgon** [comment]
Not necessary it works fine for me in latest nextjs. 
Debugging experience is awesome in VSCode and is ok in Nvim (I use nvim). 

The difference is that in nvim you have to start the nextjs server in another terminal and in vscode you can do that but also launch a debugging session from the editor which is nice but not enough for me to switch editor : )

So here is how I do debugging in next:

Start server with `NODE_OPTIONS='--inspect'` and then attach to that debugging session from nvim.

To make my life easier I have a `dev:debug` in my `package.json`

``` json
{
  "scripts": {
    "dev": "next dev",
    "dev:debug": "NODE_OPTIONS='--inspect' next dev",
  }
}
```
That's all now I can see the  breakpoints in my nvim. 
<img width="1552" alt="image" src="https://github.com/vercel/next.js/assets/49499/11323378-1f53-46cc-8ec7-df6122446363">

Beautiful!

All the problem was that nextjs debugging was broken and was fixed here https://github.com/vercel/next.js/pull/51467

**Comment by bengotow** [comment]
In case anyone stumbles across this (it's the #1 google result for the devtools warning). I was seeing nothing in my performance traces, and it was because my app was spawning subprocesses to handle requests via `throng` when NODE_ENV=production. The profiler connects to the root nodejs process, but that process is not actually doing much.

For example if you have code like this, you need to skip `throng` and run as a single process:

```
if (process.env.NODE_ENV === 'development') {
  void start(1);
} else {
  void throng({
    workers: WORKERS,
    lifetime: Infinity,
    start: start,
  });
}

```
